elastic/kibana
# [Cases] Persist the cases table state on the URL

## Summary

This PR persists the whole state of the cases table in the URL. Specifically:

- The format of the URL changed. The Rison format will be used which is widely used in Kibana.
- The old format will be supported to support BWC.
- All filters are persisted in the URL including custom fields.
- The URL is the source of truth and it takes precedence over everything.
- The state is also persisted in the local storage. The state from the local storage is loaded only when the URL is empty.
- The state in the local storage is stored under a new key called `cases.list.state`. The old key will not be taken into account.
- Navigate through pages and back to the cases table do not lose the state of the table.
- 

## Testing

- Test that the filtering is working as expected with all possible filters and values.
- Test that the URL is updating correctly as you change the filters and the pagination.
- Test that submitting a URL takes priority over the state persisted in the local storage.
- Test that submitting a URL changes the filters and the pagination as expected.
- Test that submitting a URL makes any hidden filters visible.
- Test that legacy URLs are working as expected.
- Put malformed values and keys in the URL and see how the application behaves.
- Test that the extra query params put by the Security solution persisted and do not affect the cases filtering.
- Test that the configuration of the filters (which ones are visible and which ones are hidden) is not affected.
- Hide all filters and put a URL that contains all filters. They should be shown.
- Remove the local storage state and check how the application behaves when you enter a URL with filters
- Ensure that when you navigate between pages your filtering is persisted.
- Ensure that the state from the local storage is loaded when you enter a URL without query parameters.
- No assignees filtering is working.
- Passing non-existing custom fields on the URL does not lead to an error.
- Paste a URL that is not the same as the state in the local storage, navigate to a case, and come back to the cases table. The filters set by the URL should be the same.

Blocked by: https://github.com/elastic/kibana/pull/176546
Flaky test runner: https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/5114, https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/5119

### Checklist

Delete any items that are not applicable to this PR.


### For maintainers


## Release notes

Persist all filter options of the cases table, including custom fields, in the URL. The filtering also persists when navigating back and forth between pages.
